Refactor terrain improvement system

esbudylin opened the pull request:
This PR introduces a new class, TerrainImprovement, which describes player-made improvements (roads, mines, irrigation, outposts...) Currently, it encapsulates the movement cost, the defense bonus (for barricades and fortresses), and the upgrade relationship (for example, the fact that the railroad is an upgrade from the regular road).

This PR groups the terrain improvements by layers
- Roads
- ResourceDevelopment - mines and irrigation
- Holdings - outpost, radar tower, fortress, barricade. 

The layer systems plays a limiting role when adding new improvements to the tile via the TileOverlays class: each tile can have only one type of improvement per layer.

Some of the Civ3 improvements doesn't fit well into this system: in the Civ3, there are irregularities with Colony (it can coexist with Fortress but not with Outpost or Radar Tower) and Airfield (it can't coexist with anything except roads). I think this can be addressed later by adding custom logic.

In the future, I want to move tile yield effects into the TerrainImprovement class. Also, it would be ideal to allow defining new improvements via scripting/JSON, but it would be tricky since it requires decoupling the improvement drawing logic from the TileOverlayLayer class.

esbudylin committed:
Refactor terrain improvement system

TomWerner commented on C7Engine/EntryPoints/CityInteractions.cs:
```diff
@@ -32,12 +32,6 @@ public static City BuildCity(int X, int Y, ID playerID, string name) {
-			// a city is build on a mine, the mine should be removed.
-			tileWithNewCity.overlays.road = true;
-			tileWithNewCity.overlays.mine = false;
-			tileWithNewCity.overlays.irrigation = false;
```
This seems to break road connections to cities when adding new roads.

esbudylin replied:
Thanks for the review! I've tested it, and it works as before. The logic for that was moved into the TileOverlays class (methods HasRoad and MovementCost). Now these methods explicitly handle the presence of a city on a tile, so there is no longer a need for adding/removing roads on city creation/destruction.

esbudylin replied:
Hm, but I really forgot about removing mines and irrigation on city creation, I'll fix that.

TomWerner replied:
Ah, the movement costs are correct, but the graphics are wrong. I think the road layer needs to be updated to have that check too. Here's what I see if I build a road next to a city - there's no connection.

![image](https://github.com/user-attachments/assets/ab29d0cb-91ae-4b1d-a14c-c2b7b27c757b)

esbudylin replied:
Thanks, I see it now!

TomWerner reviewed (changes requested):
I think there are some issues with city creation/destruction here, otherwise looks good

esbudylin committed:
Address review comments

- Remove terrain improvements (irrigation, mines) and
overlays (jungle, forest) on city creation

- Fix visual disconnect between a city and the road network

esbudylin commented:
@TomWerner I pushed the commit with fixes, let me know what you think!

TomWerner commented on C7/Map/TileOverlayLayer.cs:
```diff
@@ -166,7 +166,7 @@ private static bool HasAnyOverlays(Tile tile) {
+			return tile.overlays.HasImprovement(TerrainImprovement.road) || tile.HasCity;
 		}
 
 		private static bool hasRailRoad(Tile tile) {
```
I think this needs `|| tile.HasCity` too.

TomWerner replied:
Oh, actually the railroad issue is movement-related too, not just visual.

esbudylin replied:
Thanks! The issue with movement from a city to a railroad is also present on the Development brunch. Should be fixed now!

TomWerner reviewed (approved):
LGTM once the railroad thing is handled!

esbudylin committed:
Fix movement cost for cities

Moving from a city to a railroad now doesn't cost movement points

esbudylin committed:
Fix visual disconnect between a city and a railroad
